### PR TITLE
Support dot_general (matmul), pad, reshape, transpose, reverse, reduce operations (sum, min, max) and more unops/binops

### PR DESCRIPTION
**Pull request by juliuskunze**

### CODE REVIEW

File: fastar/update_rules.py [hunk 1]
```diff
@@ -21,66 +22,147 @@ def take(arr, idxs):
...
+                            for i, s in enumerate(outmask.shape)]
+
+            outmask[lower_slice] = True
+            outmask[higher_slice] = True
+
+        return outval, outmask
+
+    output_mask = onp.ones(tuple(s + lo + hi for s, (lo, hi, _) in zip(a.shape, padding_config)),
+                           dtype=bool)
+    output_mask[tuple([slice(lo, -hi, None) for lo, hi, _ in padding_config])] = a_mask
```
Comment (j-towns) [inline fastar/update_rules.py]: Can’t the new output mask be computed easily by calling pad on the input mask, with the pad value set to True or False depending on the input pad value’s mask?

File: fastar/update_rules.py [hunk 2]
```diff
@@ -21,66 +22,147 @@ def take(arr, idxs):
...
+
+    def input_slices_from_output_slice(output_slice):
+        return (tuple(slice(s.start - lo, s.stop - lo, None)
+                      for s, (lo, _, _) in zip(output_slice, padding_config)), ())
+
+    return sliceableop_update(
```
Comment (j-towns) [inline fastar/update_rules.py]: I can’t see what sliceableop_update is buying us here. Pad doesn’t fit into the pattern of ops where you get the subsections of the outputs by calling pad on subsections of the inputs.

File: fastar/update_rules.py [hunk 3]
```diff
@@ -21,66 +22,147 @@ def take(arr, idxs):
...
-    outval = lax.full(abstract_out.shape, 0, abstract_out.dtype)
+    outval = lax.full(abstract_out.shape, init_value, abstract_out.dtype)
     outmask = onp.zeros(abstract_out.shape, bool)
     return outval, outmask
 
+
+def sliceableop_update(func, old_out, *args, output_mask, input_slices_from_output_slice,
+                       init_output=_init_output, sliced_output_from_sliced_inputs=None,
+                       init_value=0,
+                       **params):
```
Comment (j-towns) [inline fastar/update_rules.py]: I’m still sceptical about whether this is a good idea. It reduces the number of lines of code, but in my opinion it also reduces readability/clarity of what’s going on by adding a layer of indirection. What do you think?

File: fastar/update_rules.py [hunk 4]
```diff
@@ -1,17 +1,18 @@
...
 from jax.abstract_arrays import make_shaped_array
+from jax.util import curry, safe_zip, safe_map, unzip2
 
 import fastar.interpreter as fa
 import fastar.primitives as primitives
 import fastar.util as util
 
 map = safe_map
 zip = safe_zip
+all = slice(None, None)
```
Comment (j-towns) [inline fastar/update_rules.py]: I think over-riding builtins should be done with care. The above two over-rides at least replace with something which is very similar.

This over-ride could confuse people reading the code quite a bit because this all has a very different meaning to the built in.

It’s annoying that there isn’t already a shorthand for this ‘empty’ slice. Could maybe call it noneslice?

File: fastar/update_rules.py [hunk 5]
```diff
@@ -21,66 +22,147 @@ def take(arr, idxs):
...
+
+        return outval, outmask
+
+    output_mask = onp.ones(tuple(s + lo + hi for s, (lo, hi, _) in zip(a.shape, padding_config)),
+                           dtype=bool)
+    output_mask[tuple([slice(lo, -hi, None) for lo, hi, _ in padding_config])] = a_mask
+
+    def input_slices_from_output_slice(output_slice):
+        return (tuple(slice(s.start - lo, s.stop - lo, None)
+                      for s, (lo, _, _) in zip(output_slice, padding_config)), ())
```
Comment (j-towns) [inline fastar/update_rules.py]: Could raise a NotImplementedError if any of the `interior_padding` values are nonzero.

**Review by j-towns — CHANGES_REQUESTED** [review]

**Comment by j-towns** [comment]
Excellent work! See comments above.

File: fastar/update_rules.py (hunk 5 above)
Comment (j-towns) [reply fastar/update_rules.py]: Also it might be slightly nicer to get the input mask from the output mask _then_ convert to slices. You can get the in mask from the outmask just by slicing away the pads.

File: fastar/update_rules.py (hunk 1 above)
Comment (juliuskunze) [reply fastar/update_rules.py]: Good point, absolutely

File: fastar/update_rules.py (hunk 4 above)
Comment (juliuskunze) [reply fastar/update_rules.py]: Good point, override is an accident. Renaming to noneslice.

File: fastar/update_rules.py (hunk 5 above)
Comment (juliuskunze) [reply fastar/update_rules.py]: > Could raise a NotImplementedError if any of the `interior_padding` values are nonzero.

Sure

Comment (juliuskunze) [reply fastar/update_rules.py]: > Also it might be slightly nicer to get the input mask from the output mask _then_ convert to slices. You can get the in mask from the outmask just by slicing away the pads.

I am not sure what you mean. Isn't the input mask is given as an argument?

File: fastar/update_rules.py (hunk 3 above)
Comment (juliuskunze) [reply fastar/update_rules.py]: I find it improves clarity by structuring and naming most relevant parts. sliceableop has a substantial amount of responsibility, e. g. mask initialization, diffing and corresponding slicing, which would be duplicated for each op class. It would further be duplicated for ops we might want to implement in the future, e. g. convs, matmul, conditional, transpose and reverse.

File: fastar/update_rules.py (hunk 2 above)
Comment (juliuskunze) [reply fastar/update_rules.py]: I made sliceableop_update applicable to all ops where any slice of the output can be calculated from one slice of each input. Using the op itself is the default though. Avoids some duplication. This generalization might not be applicable to lots of ops, so I agree, maybe it's not worth it.

File: fastar/update_rules.py (hunk 5 above)
Comment (j-towns) [reply fastar/update_rules.py]: Sorry, wasn't very clear there. I mean you could do things in this order:
1. Compute the new outmask by calling pad on the inmask as described above
2. Work out a mask for the elements of the output that need to be updated by doing `new_outmask & ~old_outmask`
3. Map that mask back to input space by 'unpadding' it
4. Convert to slices with mask_to_slices

File: fastar/update_rules.py (hunk 3 above)
Comment (j-towns) [reply fastar/update_rules.py]: I agree it would be good to get rid of the duplicated output initialization, which is going to be the same for _every_ op, i.e. this kind of line:
```
outval, old_outmask = old_out[0] if old_out else _init_out(func, a)
```
However I think we could have a lightweight function (maybe a decorator), that only does that, or handle it in the interpreter instead of passing `None` on [this line]( https://github.com/j-towns/fastar/blob/master/fastar/interpreter.py#L55).

Comment (j-towns) [reply fastar/update_rules.py]: Maybe we could have another small, separate, helper function which runs the for loop over slices, which I think we also need to do for every op. I.e. two small helper functions instead of one big one.

Comment (j-towns) [reply fastar/update_rules.py]: The gain from these utility functions is limited at the moment though since we're covering so few ops.

**Comment by j-towns** [comment]
Do you mind using a max line length of 79 or 80? I often work on my laptop and having narrow files improves the readability for me quite a lot.

I don't mind breaking that convention occasionally.

**Comment by j-towns** [comment]
For example I don't mind if test one liners like the ones in https://github.com/j-towns/fastar/blob/master/tests/test_update_rules.py exceed 80 chars.

**Comment by juliuskunze** [comment]
Ok, will do 79

File: fastar/update_rules.py (hunk 5 above)
Comment (juliuskunze) [reply fastar/update_rules.py]: Then you would need to calculate a corresponding output slice for each input slice (instead of vice versa). Why would you prefer this?

Comment (j-towns) [reply fastar/update_rules.py]: That's a good point, sorry, thought we could get away without computing the output slices but we need them for updating the output value.

File: fastar/update_rules.py (hunk 2 above)
Comment (juliuskunze) [reply fastar/update_rules.py]: Just implemented support for pad with uninitialized padding_value. This required a different structure from all other operators anyways, so it's now independent from sliceableop_update.

File: fastar/update_rules.py (hunk 3 above)
Comment (juliuskunze) [reply fastar/update_rules.py]: Good points. Based on your feedback, I simplified `sliceableop_update`:

- As you suggested, I moved `_init_out` into the interpreter (replacing `None` in `firstpass`). This saves us from calling the same helper function for every op. 
- Additionally, I made the now quite different case of `pad` independent (see above), saving us 3 arguments.

Since that made it much more readable and it has seven use cases (all currently implemented op classes except pad, namely unop and binop, dot, dot_general, reduce, transpose, reverse), I think it might be a more useful abstraction now. What do you think?

File: fastar/update_rules.py [hunk 6]
```diff
@@ -21,66 +20,218 @@ def take(arr, idxs):
...
+def transpose_update(old_out, a, permutation):
+    a, a_mask = a
+
+    def inversely_permute_slice(s):
+        r = [slice(None)] * len(s)
+        for i, d in enumerate(permutation):
+            r[d] = s[i]
+        return tuple(r)
+
+    return sliceableop_update(
```
Comment (j-towns) [inline fastar/update_rules.py]: I think for transpose it would (probably) be faster, or roughly as fast, to simply run `transpose(a, permutation)`, rather than updating individual elements.
